diff --git a/init/main.c b/init/main.c
index 63b3eafd..e11c96b2 100644
--- a/init/main.c
+++ b/init/main.c
@@ -631,6 +631,9 @@ asmlinkage __visible void __init start_kernel(void)
 			   set_init_arg);
 
 #ifdef CONFIG_TIMA_RKP
+#ifdef CONFIG_KVM
+#error "RKP and KVM cannot coexist!"
+#endif
 #ifdef CONFIG_KNOX_KAP
 	if (boot_mode_security)
 		vmm_init();
@@ -651,6 +654,10 @@ asmlinkage __visible void __init start_kernel(void)
 	sort_main_extable();
 	trap_init();
 	mm_init();
+#ifdef CONFIG_KVM
+    void preinit_hyp_mode(void);
+    preinit_hyp_mode();
+#endif
 
 	/*
 	 * Set up the scheduler prior starting any interrupts (such as the
